File: Tests/ButterflyNet.Tests/IllegalScopingTests.cs
using NUnit.Framework;

namespace ButterflyNet.Parser.Tests {
	[TestFixture]
	public class IllegalScopingTests : WikiToHtmlTest {

		[Test]
		public void Should_not_create_paragraph_if_inline_scope_is_not_closed() {
			AssertWithNoRegardForLineBreaks(Parser.ParseAndReturn("__foo\n\n\nbar__"), "<p><strong>foobar</strong></p>");
		}

		[Test]
		[ExpectedException(typeof(ParseException), ExpectedMessage = "Scopes that need to be manually closed were not closed")]
		public void Should_throw_if_manually_closing_scope_is_not_closed() {
			Parser.Parse("__foo");
		}

		[Test]
		[ExpectedException(typeof(ParseException), ExpectedMessage = "Cannot close table row until all containing scopes are closed")]
		public void Should_throw_if_containing_scopes_in_table_row_are_not_closed() {
			Parser.Parse("|{ __foo }|");
		}

		[Test]
		[ExpectedException(typeof(ParseException), ExpectedMessage = "Cannot close table cell until all containing scopes are closed")]
		public void Should_throw_if_containg_scopes_in_table_cell_are_not_closed() {
			Parser.Parse("| __foo | bar |");
		}

	}
}

File: src/js/tests/IllegalScopingTests.js
var illegalScopingTests = function() {
	var parser;
	
	return {
		setup: function() {
			parser = createParser();
		},
		
		tearDown: function() {
			parser = null;
		},
		
		test: function Illegal_scoping_tests() {
			return [
				function Should_not_create_paragraph_if_inline_scope_is_not_closed() {
					Assert.that(trimLf(parser.parseAndReturn("__foo\n\n\nbar__")), Is.equalTo("<p><strong>foobar</strong></p>"));
				},
				
				function Should_throw_if_manually_closing_scope_is_not_closed() {
					Assert.willThrow(new Butterfly.ParseException("The following scope must be manually closed: strong"));
					parser.parse("__foo");
				},
				
				function Should_throw_if_containing_scopes_in_table_row_are_not_closed() {
					Assert.willThrow(new Butterfly.ParseException("Cannot close table row until all containing scopes are closed"));
					parser.parse("|{ __foo }|");
				}
			];
		}
	};
}();

Jarvis.run(illegalScopingTests);